---
title: rawR - Technical Note, Journal of Proteome Research, Second Biennial Special
  Issue on Software Tools and Resources in February 2021
author: "Tobias Kockmann^1‡^ & Christian Panse^1,2‡^"
date: "`r Sys.Date()`"
bibliography: rawR.bib
output:
  pdf_document: default
  html_document:
    df_print: paged
    citation_package: natbib
vignette: |
  %\usepackage[utf8]{inputenc}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::knitr} %\VignetteIndexEntry{rawR JPR Technical Note}
  %\VignetteEngine{knitr::rmarkdown}
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

^1^Functional Genomics Center Zurich, ETH Zurich / University of Zurich,
Winterthurerstrasse 190, 8057 Zurich, Switzerland

^2^Swiss Institute of Bioinformatics, Quartier Sorge - Batiment Amphipole,
1015 Lausanne, Switzerland

## Abstract

The Bioconductor project has shown that the R statistical environment is a
highly valuable tool for genomics data analysis
[@Huber2015], but with respect to proteomics
we are still missing low level infrastructure to enable performant and robust
analysis workflows in R. Fundamentally important are libraries that provide
raw data access. Our R package rawDiag [@Trachsel2018]
has provided the proof-of-principle how access to mass spectromerty raw files
can be realized by wrapping vendor-provided APIs. Our novel package rawR now
aims for complete, OS independent access to all spectral, chromatographic and
meta data logged in Thermo Fisher Scientific raw files. rawR interoperates with
other R packages for proteomics data visualisation (protViz) and will ...

## Keywords

computational mass spectrometry, software, R package

## Introduction

...


## Implementation

### C# code

### R

Mass spectrometry (MS) uses two basic building blocks (data items): 1. mass spectrum, 2. chromatogram. All mass spectra are recorded by scanning detectors (mass analyzers) that log signal intensities for ranges of mass to charge ratios (m/z), also referred to as position. These recordings can be of continuous nature so called profile data, or appear centroided in case discrete information (tuples of position and intensity values) is sufficient. This heavily compacted data structure is often called a peak list. In addition to signal intensities, peak list can also cover additional attributes like peak resolution, charge, or local noise around the peak. In short, the additional attributes further described the nature of the original profile signal, or help to group peak lists with respect to their molecular nature or processing history. A well-known example is the assignment of peaks to peak groups that constitute isotope patterns (M, M+1, M+2, ...). Chromatograms come in different flavours, but are always signal intensity values as a function of time. Signal intensities can be point estimates from scanning detectors or plain intensities from non scaning detectors (e.g. UV trace). Point estimates can be defined in different ways by for instance summing all signals of a given spectrum (total ion chromatogram or TIC), or by extracting signal around an expected value (extracted ion chromatogram = XIC), or by using the maximum signal (base peak chromatogram = BPC). On top, chromatograms can be computed from pre-filtered lists of scans. A total ion chromatogram (TIC) for instance is typically generated by iterating over all MS1-level scans. 

We therefore decided to implemented objects in R that closely resemble these basic building blocks. The spectrum object is technically a nested list (S3 class system). Apart from the peak list the spectrum object also stores meta information on how the scan was recorded and how it relates to other scans in the raw file. This meta information is grouped for better transparency. The chromatogram object ...


### Example data

<!---
TODO(cp): add to man page
The binary example file sample.raw contains 574 fourier-transformed orbi trap spectra (FTMS) recorded on a Thermo Fisher Scientific Q Exactive HF-X. The mass spectrometer was operated in line with a nano electrospray source (NSI) in positive mode (+). All spectra were written to disk after applying centroiding (c) and lock mass correction. Additional raw data for demonstration and extended testing is available through the [tartare package](https://bioconductor.org/packages/tartare/). Lions love raw meat!
-->

## Results



|function|description|
|--------|-----------|
|readFileHeader|extracts some meta information from a given rawfile.|
|readIndex|read scan index|
|readSpectrum|reads scan information, e.g., charge, mZ, or intensity of a given set of scan numbers using a dot net interface and the ThermoFisher NewRawFileReader libraries.|
|readChromatogram|extracts Chromatogram|


### Use Case I - Analyzing orbi trap spectra
<!---
scp fgcz-r-035.uzh.ch:/export/lv_iduzh06/projects/p1000/Proteomics/QEXACTIVEHF_2/tobiasko_20181113/20181113_010_autoQC01.raw ~/Downloads/
-->

plotting

using biognosys iRT peptides

```{r plot.rawRspectrum, fig.cap="graphs a tandem mass spectrum of the LGGNEQVTR peptide. The vertical grey lines indicate the in-silico computed y-ions of the peptide sequence.", error=TRUE}
# http://fgcz-ms.uzh.ch/~cpanse/20181113_010_autoQC01.raw
# MD5 (20181113_010_autoQC01.raw) = a1f5df9627cf9e0d51ec1906776957ab
rawfile <- file.path(Sys.getenv('HOME'),
                     "Downloads", "20181113_010_autoQC01.raw")

library(rawR)

S <- rawR::readSpectrum(rawfile, scan=c(9594, 11113, 11884, 12788, 12677, 13204,
                                         13868, 14551, 16136, 17193, 17612))
plot(S[[1]], centroid=TRUE)


# derived using  https://CRAN.R-project.org/package=protViz
yion <- c(175.1190, 276.1666, 375.2350, 503.2936, 632.3362, 746.3791, 803.4006,
       860.4221)
abline(v = yion, col='#DDDDDD88', lwd=5)
axis(3, yion, paste0("y",seq(1,length(yion))))
```

other recent application examples are descibed in [@Panse2020], [@protViz]
and [@Gehrig2020].


### Use Case II - iRT regression

By applying linear regression one can convert observed peptide retention times (RTs) into dimensionless scores termed iRT values (iRTs) and *vice versa* [@Escher2012]. In addition, fitted iRT regression models provide highly valuable information about LC-MS run performance. In this example we show how easy it is to perform iRT regression in R by just using the raw measurement data, our package rawR, and well known base R functions supporting linear modeling. The first step is to estimate the empirical RTs of a peptide set with known iRT scores from a raw file. In the simplest case, this is achieved by computing an extracted ion chromatogram (XIC) for iRT peptide ions, given they were spiked into the sample matrix prior to data acquisition. Code chunk X demonstrates how the function `readChromatogram()` is called on the R command line to return a `rawRchromatogram` object. This object is printed for visual inspection. 

```{r plot.rawRchromatogram, fig.cap="XICs iRT peptides", error=TRUE}
iRT <- c(487.2571, 547.2984, 622.8539, 636.8695, 644.8230, 669.8384, 683.8282,
            683.8541, 699.3388, 726.8361, 776.9301)

names(iRT) <- c("LGGNEQVTR", "YILAGVENSK", "GTFIIDPGGVIR", "GTFIIDPAAVIR",
                 "GAGSSEPVTGLDAK", "TPVISGGPYEYR", "VEATFGVDESNAK",
                 "TPVITGAPYEYR", "DGLDAASYYAPVR", "ADVTPADFSEWSK",
                 "LFLQFGAQGSPFLK")

C <- readChromatogram(rawfile, mass = iRT, tol = 10, type = "xic", filter = "ms")
plot(C)
```

For regression, we now extract the RTs at the maximum of intensity traces stored in the chromatogram object and fit a linear model of the form $$rt = a + b\cdot score$$.

```{r iRTscoreFit, error=TRUE}
score <- c(-24.92, 19.79, 70.52, 87.23, 0, 28.71, 12.39, 33.38, 42.26, 54.62, 100)
rt <- sapply(C, function(x) x$times[which.max(x$intensities)[1]])
fit <- lm(rt ~ score)
```



```{r iRTscoreFitPlot, echo=FALSE, error=TRUE, fig.cap="graphs the retention time verus the retention time scores."}
plot(rt ~ score,
     ylab = 'Retention time [min]',
     xlab = "iRT score",
     pch=16,frame.plot = FALSE)
abline(fit, col = 'grey')
legend("topleft", legend = paste("regression line: ", "rt =",
                                 format(coef(fit)[1], digits = 4), " + ",
                                 format(coef(fit)[2], digits = 2), "score",
                                 "\nR2: ", format(summary(fit)$r.squared, digits = 2)),
       bty = "n", cex = 0.75)
text(score, rt,  iRT,pos=1,cex=0.5)
```



## Conclusions

... next 

- [@Schmidt2020] 
- [@biocspectra]





## Author information

### Corresponding author

Tobias Kockmann  
E-mail: <tobias.kockmann@fgcz.ethz.ch>

### ORCID

Christian Panse <https://orcid.org/0000-0003-1975-3064>  
Tobias Kockmann <https://orcid.org/0000-0002-1847-885X>  

### Author contributions

The manuscript was written through contributions of all authors. All authors have given approval to the final version of the manuscript. ‡These authors contributed equally.


## Funding Sources

Any funds used to support the research of the manuscript should be placed here (per journal style).


## Acknowledgements

We thank Lilly van de Venn for designing the rawR package logo.


## Abbreveations

MS, mass spectromerty; TIC, total ion chromatogram;
XIC, extracted ion chromatogram; fourier-transformed mass spectrum (FTMS);
nano spray ionisation, NSI;


# References



