---
title: "The rawrr R package – Direct access to orbitrap data and beyond"
author: "Tobias Kockmann^1‡\\*^ & Christian Panse^1,2‡^"
date: "`r Sys.time()`"
output:
  pdf_document: 
    keep_tex: yes
    latex_engine: xelatex
  citation_package: natbib
bibliography: rawrr.bib
csl: american-chemical-society.csl
vignette: |
  %\usepackage[utf8]{inputenc}
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{rawrr - Technical Note}
  %\VignetteEngine{knitr::rmarkdown}
urlcolor: blue
header-includes:
  - \usepackage{setspace}\doublespacing
keywords: computational mass spectrometry, software, R package
abstract: |
  The Bioconductor project[@Huber2015] has shown that the `R` statistical environment is a highly valuable tool for genomics data analysis, but with respect to proteomics, we are still missing low-level infrastructure to enable performant and robust analysis workflows in `R`. Fundamentally important are libraries that provide raw data access. Our `R` package `rawDiag`[@Trachsel2018] has provided the proof-of-principle how access to mass spectrometry raw files can be realized by wrapping a vendor-provided advanced programming interface (API), but rather focused on metadata analysis and visualization. Our novel package `rawrr` now provides complete, OS independent access to all spectral data logged in Thermo Fisher Scientific raw files. In this technical note, we present implementation details and describe the main functionality provided by the `rawrr` package. In addition, we report two use cases inspired by real-world research tasks that demonstrate the application of the package. The raw data used for demonstration purposes was deposited as MassIVE dataset [MSV000086542](https://massive.ucsd.edu/ProteoSAFe/dataset.jsp?task=575538e190e84cbfbf6c17aa1219e403). **Availability**: <https://github.com/fgcz/rawrr> **Keywords**: computational mass spectrometry, software, R package
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

1 Functional Genomics Center Zurich, ETH Zurich / University of Zurich,
Winterthurerstrasse 190, 8057 Zurich, Switzerland

2 Swiss Institute of Bioinformatics, Quartier Sorge - Batiment Amphipole,
1015 Lausanne, Switzerland

\* Corresponding author

‡ Equal contributing author




<!--
Deposition of Mass spectrometry data

For papers which were initially submitted without depositing MS data into a PUBLIC repository (such as ProteomeXchange, Massive, or Chorus),  deposit the data PRIOR to submitting a revision, and include dataset identifier(s) in the Abstract of the revised paper to show compliance.  This is a critical point for publication transparency since it allows readers to reproduce results claimed in the manuscript.
NOTE:   MS data provided on author-managed links is NOT acceptable.
-->


# Introduction

Mass spectrometry-based proteomics and metabolomics are the preferred technology to study the protein and metabolite landscape of complex biological systems. The orbitrap mass analyzer is one of the key innovations that propelled the field by providing HRAM data on a chromatographic time scale. Driven by the need to analyze the resulting LC-MS data, several specialized software tools have been developed in the last decade. In the academic environment, [MaxQuant](https://maxquant.org/)[@Cox2008] and [Skyline](https://skyline.ms/project/home/begin.view?)[@MacLean2010] are beyond the most popular ones. These software tools usually offer GUIs that control running predefined analysis templates/workflows, including free parameters that need to be defined by the user. In parallel, projects like [OpenMS](https://www.openms.de/)[@Rst2016] or [pyteomics](https://github.com/levitsky/pyteomics)[@Goloborodko2013] developed, but chose a fundamentally different approach. They aim at providing software libraries bound to specific programming languages like `C++` or `Python`. These naturally offer greater analytical flexibility but require programming skills from the end-user and have therefore not reached the popularity of their GUI counterparts. Proteomics and metabolomics specific libraries have also been developed for the [`R`](https://www.r-project.org/) statistical environment, but these mainly support high-level statistical analysis once the raw measurement data has undergone extensive preprocessing and aggregation by external software tools (often the GUI-based once listed above). A typical example is the `R` package [MSstats](http://msstats.org/)[@Choi2014] for the statistical analysis of LC-MS experiments with complex designs or [MSqRob](https://github.com/statOmics/MSqRob)[@Goeminne2015]. MSstats can process MaxQuant or Skyline output and creates protein/peptide level estimates whether the biological system shows statistically significant regulation. In a nutshell, these tools provide statistical postprocessing. Libraries that support working with the spectral data in `R` also exist, for instance, the Bioconductor package [MSnbase](https://www.bioconductor.org/packages/release/bioc/html/MSnbase.html)[@Gatto2011], but require conversion of raw data to exchange formats like [mzML](http://www.psidev.info/mzML). These conversions are primarily supported by the [ProteoWizard](http://proteowizard.sourceforge.net/)[@Chambers2012] project and its software tool `MSconvert`.

We strongly believe that a library providing raw data reading functionally would finally close the gap and facilitate modular end-to-end analysis pipeline development in `R`. This could be of special interest to research environments/projects dealing with either big data analytics or scientists interested in code prototyping without formal computer science education.  Another key aspect regarding multi-omics integration of proteomics and metabolomics data is the fact that high-throughput genomic data analysis is already done mostly in `R`.
<!-- So proteomics and metabolomics could finally "join the party" !-->
This is primarily due to the [Bioconductor project](https://www.bioconductor.org/)[@Huber2015] that currently provides >1900 open-source software packages, training & teaching, and a very active user & developer community. Having these thoughts in mind, we decide to implement our R package `rawrr`. `rawrr` utilizes a vendor-provided API to access spectral data logged in proprietary raw files. These binary files are written by all orbitrap mass spectrometers, unlocking an incredible amount of the global LC-MS data, also stored in public repositories like [ProteomeExchange](http://www.proteomexchange.org/). This manuscript presents a first package version/release and shows case its usage for bottom-up proteomics data analysis.

# Implementation

Our implementation consists of two language layers, the top `R` layer and the hidden `C#` layer. Specifically `R` functions requesting access to data stored in binary raw files (reader family functions listed in Table 1) invoke compiled `C#` wrapper methods using a system call. Calling a wrapper method typically results in the execution of methods defined in the `RawFileReader` dynamic link library provided by Thermo Fisher. Our precompiled wrapper methods are bundled in the `rawrr` executable file and shipped with the  released `R` package. Running `rawrr.exe` requires the [mono](https://www.mono-project.com/) environment on none Microsoft operating systems. Mono is a cross platform, open source .NET framework. On Microsoft Windows the Microsoft .NET framework is typically already installed and sufficient. Our package also contains the `C#` source code (rawrr.cs), hopefully allowing other developers to follow and improve our code. In order to return extracted data back to the `R` layer we use file I/O. More specifically, the extracted information is written to a temporary location on the harddrive, read back into memory and parsed into `R` objects.

<!--
The implementation section is missing a detailed description on how the classes “spectrum” and “chromatogram” look like. As these are the “only two non-standard objects” the reader should be informed about which properties they contain (without having to investigate the package documentation). It might make sense to attach the documentation as supporting material, although it is a current snapshot and will change with further development of the package. 
-->

Since mass spectrometry typically uses two basic data items, the mass spectrum and the mass chromatogram, we decided to implement corresponding objects following `R`'s `S3` OOP system [@newS] named `rawrrspectrum` and `rawrrchromatogram`. The package provides functions to create and validate class instances (objects). However, typically instances are generated by the reader family of functions listed in Table 1. and returned as lists, so called sets (`rawrrspectrumSet`, `rawrrchromatogramSet`). The class constructors (`rawrrspectrum` and `rawrrchromatogram`) primarily exist for (unit) testing purposes. The internal architecture of `rawrrspectrum` is a linear list of vectors. List item names are typically transferred as returned by the `RawFileReader` and minimal type conversion is applied. The data portion of the `rawrrchromatogram` object is as well a simple linear list. We also implemented basic generics for printing and plotting of objects that we used for visualization throughout this manuscript. To minimize dependencies, we choose to stick to `base R`.

<!--
Comment: Maybe to basic for JPR, since MS community is familiar with the definition of spectra and chromatogram

#### Spectra
All mass spectra are recorded by scanning detectors (mass analyzers) that log signal intensities for ranges of mass to charge ratios (m/z), also referred to as position. These recordings can be of continuous nature, so-called profile data (p), or appear centroided (c) in case discrete information (tuples of position and intensity values) are sufficient. This heavily compacted data structure is often called a peak list. In addition to signal intensities, a peak list can also cover additional peak attributes like peak resolution (R), charge (z), or local noise estimates. In short, the additional attributes further described the nature of the original profile signal or help to group peak lists with respect to their molecular nature or processing history. A well-known example is the assignment of peaks to peak groups that constitute isotope patterns (M, M+1, M+2, ...). 

#### Chromatograms
Chromatograms come in different flavors but are always signal intensity values as a function of time. Signal intensities can be point estimates from scanning detectors or plain intensities from non-scanning detectors, e.g., UV trace. Scanning detector (mass analyzers) point estimates can be defined in different ways by, for instance, summing all signals of a given spectrum (total ion chromatogram or TIC), or by extracting signal around an expected value (extracted ion chromatogram = XIC), or by using the maximum signal contained in a spectrum (base peak chromatogram = BPC). On top, chromatograms can be computed from pre-filtered lists of scans. A total ion chromatogram (TIC), for instance, is typically generated by iterating over all MS1-level scans.
-->

### Example data
The example file `20181113_010_autoQC01.raw` used throughout this manuscript contains Fourier-transformed orbitrap spectra (FTMS) recorded on a Thermo Fisher Scientific Q Exactive HF in positive mode (+). The mass spectrometer was operated in line with a nano UPLC and a nano electrospray source (NSI). MS2 spectra were generated by HCD fragmentation at normalized collision energy (NCE) of 27. All spectra were written to file after applying centroiding (c) and lock mass correction. The analyzed sample consisted of the iRT peptide mix (Biognosys) in a tryptic BSA digest (NEB) and was separated applying a 20 min linear gradient on C18 reversed-phase material at a constant flow rate of 300 nl/min. The file is part of the MassIVE dataset [MSV000086542](https://massive.ucsd.edu/ProteoSAFe/dataset.jsp?accession=MSV000086542) [@MSV000086542].

Additional raw data for demonstration and extended testing is available through the Bioconductor data package [tartare](https://bioconductor.org/packages/tartare/) [@tartare].

<!---
TODO(cp): add to man page
The file sample.raw contains 574 Fourier-transformed orbitrap spectra (FTMS) recorded on a Thermo Fisher Scientific Q Exactive HF-X. The mass spectrometer was operated in line with a nano electrospray source (NSI) in positive mode (+). All spectra were written to disk after applying centroiding (c) and lock mass correction. Additional raw data for demonstration and extended testing is available through the [tartare package](https://bioconductor.org/packages/tartare/) [@tartare]. Lions love raw meat!
-->

# Results

The following sections are inspired by real-world research/infrastructure projects but have been stripped down to the bare scientific essentials to put more weight on the software application. We display source code in grey-shaded boxes, including syntax highlights. Corresponding `R` command line output starts with `##` and is shown directly below the code fragment that triggered the output. All figures are generated using the generic plotting functions of the package.

|Function Name     |Description                                             |Return value            |
|:-----------------|--------------------------------------------------------|:-----------------------|
|`readFileHeader`  |Reads meta information from a rawfile header.           |`list`                  |
|`readIndex`       |Reads scan index from a rawfile.                        |`data.frame`            |
|`readSpectrum`    |Reads spectral data from a rawfile                      |`rawrrSpectrum(Set)`    |
|`readChromatogram`|Extracts chromatograms (TIC, BPC, or XIC) from a rawfile|`rawrrChromatogram(Set)`|

Table: lists the most important `rawrr` package functions connected to reading functionality. More details can be found in the package documentation.

## Use Case I - Analyzing Orbitrap Spectra

```{r readFileHeader, echo=FALSE, message=FALSE}
library(rawrr)
rawfile <- rawrr::sampleFilePath()
H <- rawrr::readFileHeader(rawfile = rawfile)
```

The orbitrap detector has been a tremendous success story in MS, since it offers high resolution, accurate mass (HRAM) data on a time scale that is compatible with chromatographic analysis (LC-MS). Therefore. it is heavily used for bottom-up proteomics, but analyzing orbitrap data in `R` has so far only been possible after raw data transformation to exchange formats like mz(X)ML. This use case shows how easy it is to work directly with the binary raw data after installing our `R` package `rawrr` that applies vendor APIs for data access. We use a complete LC-MS run recorded on a `r H$"Instrument model"` for demonstration purposes. The `r H$"Time range"[2]` min run resulted in `r H$"Number of scans"` scans that were written to the file. Already typesetting the above lines uses `rawrr` functionality, since the instrument model, the time range of data acquisition, and the number of scans is extracted from the binary file header (Note: This manuscript was written in `R markdown` and combines `R` code with narration). The respective function is called `readFileHeader()` and returns a simple `R` object of type `list` (see Table 1).

Individual scans or collection (sets) of scans can be read by the function `readSpectrum()` which returns a `rawrrspectrum` object or `rawrrspectrumSet`. Our package also provides generics for printing and plotting these objects. The following code chunk depicts how a set of scans is read from the rawfile and the corresponding Figure 1 shows the resulting plot for scan `9594`:

```{r plotrawrrspectrum, fig.cap="Plot of scan number 9594 showing a centroided tandem mass spectrum of the iRT peptide precursor LGGNEQVTR++ in positive mode. The scan was acquired on an orbitrap detector incl. lock mass correction and using a transient of 64 ms (equal to a resolving power of 30'000 at 200 m/z) and injection of 100'000 charges (AGC target). Peak attributes like m/z, charge (z), and resolution (R) are shown above the peaks.", error=TRUE}
library(rawrr)
# ftp://massive.ucsd.edu/MSV000086542/raw/20181113_010_autoQC01.raw
# MD5: (20181113_010_autoQC01.raw) = a1f5df9627cf9e0d51ec1906776957ab
rawfile <- file.path(Sys.getenv('HOME'), "Downloads", "20181113_010_autoQC01.raw")
## scan numbers (PSMs) derived from DB search
i <- c(9594, 11113, 11884, 12788, 12677, 13204, 13868, 14551, 16136, 17193, 17612)
S <- rawrr::readSpectrum(rawfile = rawfile, scan = i)
class(S[[1]])
summary(S[[1]]) #comment this line if necessary to save space

plot(S[[1]], centroid=TRUE)
```

The plot shows typical orbitrap peak attributes like resolution (R) and charge (z) above the most intense peaks when centroided data is available and selected. Centroided data also makes it possible to graph spectra using signal-to-noise as response value. This is potentially interesting since orbitrap detectors follow $$S/N \sim charges \cdot \sqrt R$$ and signal-to-noise makes judging the signal quantity more intuitive than using arbitrary signal intensity units. Figure 2 shows that all fragment ion signals are several ten or even a hundred folds above the local noise estimate (Note: The spectrum shown in Figure 2 is decorated with y-ions calculated for LGGNEQVTR++ using the ProtViz package [https://CRAN.R-project.org/package=protViz].).

```{r plotSN, eval = TRUE, fig.cap = "Spectrum plot using Signal/Noise option. The vertical grey lines indicate the *in-silico* computed y-ions of the peptide precusor LGGNEQVTR++ as calculated by the [protViz]( https://CRAN.R-project.org/package=protViz) package [@protViz]."}
plot(S[[1]], centroid=TRUE, SN = TRUE)

## S/N threshold indicator
abline(h = 5, lty = 2, col = "blue")

## decorate plot with y-ion series of target peptide
yIonSeries <- c(175.1190, 276.1666, 375.2350, 503.2936, 632.3362, 746.3791,
                803.4006, 860.4221)

names(yIonSeries) <- paste0("y", seq(1, length(yIonSeries)))
abline(v = yIonSeries, col='#DDDDDD88', lwd=5)
axis(3, yIonSeries, names(yIonSeries))
```


More sophisticated analysis workflows applying `rawrr` functionality have been described in [@Gehrig2020]. In short, marker ions found in HCD MS2 spectra for ADP-ribosylated peptides were annotated and cross-compared at different collision energies. Be aware that such things are convenient since the `rawrrspectrum` object provides easy access to normalized and absolute collision energies. A small molecule application using UVPD dissociation is described in [@Panse2020]. 


## Use Case II - iRT Regression for System Suitability Monitoring

By applying linear regression, one can convert observed peptide retention times (RTs) into dimensionless scores termed iRT values and *vice versa* [@Escher2012]. This can be used for retention time calibration/prediction. In addition, fitted iRT regression models provide highly valuable information about LC-MS run performance. This example shows how easy it is to perform iRT regression in `R` by just using the raw measurement data, our package `rawrr`, and well known `base R` functions supporting linear modeling. To get a first impression of the data we calculate a TIC using the `readChromatogram()` function. Plotting the TIC shows chromatographic peaks between 15 and 28 min that could be of peptidic origin. Of note, there is also a `type = "bpc"` option if you prefer a BPC:

```{r TIC, fig.cap="TIC"}
plot(rawrr::readChromatogram(rawfile = rawfile, type = "tic"))
```

```{r BPC, fig.cap="BPC", eval=FALSE}
plot(rawrr::readChromatogram(rawfile = rawfile, type = "bpc"))
```

The initial step of iRT regression is to estimate the empirical RTs of a peptide set with known iRT scores. In the simplest case, this is achieved by computing an extracted ion chromatogram (XIC) for iRT peptide precursors, given they were spiked into the sample matrix prior to data acquisition. Fortunately, our example data is iRT peptides in a tryptic digest of BSA. The code chunk below demonstrates how the function `readChromatogram()` is called on the R command line to return a `rawrrchromatogramSet` object of the type `xic`. This object is plotted for visual inspection.

```{r plotrawrrchromatogram, fig.cap="XICs for iRT peptides precursors. Each XIC was calculated using a tolerance of 10 ppm around the target mass and using only MS1 scans.", error=TRUE}
iRTmZ <- c(487.2571, 547.2984, 622.8539, 636.8695, 644.8230, 669.8384, 683.8282,
            683.8541, 699.3388, 726.8361, 776.9301)

names(iRTmZ) <- c("LGGNEQVTR", "YILAGVENSK", "GTFIIDPGGVIR", "GTFIIDPAAVIR",
                 "GAGSSEPVTGLDAK", "TPVISGGPYEYR", "VEATFGVDESNAK",
                 "TPVITGAPYEYR", "DGLDAASYYAPVR", "ADVTPADFSEWSK",
                 "LFLQFGAQGSPFLK")

C <- rawrr::readChromatogram(rawfile, mass = iRTmZ, tol = 10, type = "xic", filter = "ms")
plot(C, diagnostic = TRUE)
```

Be reminded that the intensity traces are not computed within `R`, for instance, by reading all scans of a raw file and subsequently iterating over a scan subset. Of note, this would be a greedy and slow performing solution. Instead, traces are directly calculated by a `C#` method (reference method code) that calls the vendor API. The API takes care of the filtering process (checks filter validity and applies the filter). On the `R` code level there is no need to know *a priori* which scans match the filter rule or implement vectorized operations (we generate multiple XICs simultaneously here). Only the API-returned output needs to be parsed into `rawrrchromatogram` objects. By changing the filter, one can easily switch between generating precursor traces and fragment ion traces. The following code chunk shows how to create fragment ion chromatograms (y6 to y8) generated from scans that target LGGNEQVTR++:

```{r fragmentIonTraces, fig.cap="XICs for LGGNEQVTR++ fragment ions y6 to y8", fig.retina=1}
plot(rawrr::readChromatogram(rawfile = rawfile,
      mass = yIonSeries[c("y6", "y7", "y8")],
      type = 'xic', tol = 10,
      filter = "FTMS + c NSI Full ms2 487.2567@hcd27.00 [100.0000-1015.0000]"))
```

It becomes obvious that our example data was actually recorded using parallel reaction monitoring (PRM) since 487.2567 was targeted in regular spaced intervals. You could confirm this by using the `readIndex()` function which returns a `data.frame` that indexes all scans found in a raw file and subsetting it for the scans of interest. The delta between consecutive scans is always 22 scans:


```{r readIndex}
Idx <- rawrr::readIndex(rawfile = rawfile)
head(subset(Idx,
  scanType == "FTMS + c NSI Full ms2 487.2567@hcd27.00 [100.0000-1015.0000]"))
```

For regression, we now extract the RTs at the maximum of the intensity traces stored in the chromatogram object and fit a linear model of the form: $$rt = a + b\cdot score$$ In theory, we could do this at the precursor or fragment ion level. For simplicity, we show only the first option.  

```{r iRTscoreFit, error=TRUE}
iRTscore <- c(-24.92, 19.79, 70.52, 87.23, 0, 28.71, 12.39, 33.38, 42.26, 54.62, 100)
rt <- sapply(C, function(x) x$times[which.max(x$intensities)[1]])
fit <- lm(rt ~ iRTscore)
```

The fitted model can then be inspected using standard procedures. The output of code chunk `iRTscoreFitPlot` in Figure 6, shows a visual inspection by plotting observed RTs as a function of iRT score together with the fitted model regression line. The corresponding R squared indicates that the RTs behave highly linear. This is expected since the iRT peptides were separated on a 20 min linear gradient from 5% buffer B to 35% buffer B using C18 reversed-phase material (the change rate is therefore 1.5% buffer B / min). The magnitude of the slope parameter (b) is a direct equivalent of this gradient change rate. The intercept (a) is equal to the predicted RT of iRT peptide `GAGSSEPVTGLDAK` since it was defined to have a zero score on the iRT scale.


```{r iRTscoreFitPlot, echo=FALSE, error=TRUE, fig.cap="iRT score fit plot with regression line."}
# iRTscoreFitPlot
plot(rt ~ iRTscore,
     ylab = 'Retention time [min]',
     xlab = "iRT score",
     pch=16,frame.plot = FALSE)
abline(fit, col = 'grey')
abline(v = 0, col = "grey", lty = 2)
legend("topleft", legend = paste("Regression line: ", "rt =",
                                 format(coef(fit)[1], digits = 4), " + ",
                                 format(coef(fit)[2], digits = 2), "score",
                                 "\nR2: ", format(summary(fit)$r.squared, digits = 2)),
       bty = "n", cex = 0.75)
text(iRTscore, rt, iRTmZ, pos=1,cex=0.5)
```

## Extension

An extended and dynamic version of the above use cases can be found at (https://fgcz-ms.uzh.ch/~cpanse/rawrr/test/functional_test.html). The web page displays spectra and iRT regression models obtained over a set of raw files recorded approximately every 12 hours on different orbitrap mass spectrometers at the FGCZ (some systems are out of service). The original purpose of these injections is automated longitudinal system suitability monitoring and quality control. We reuse the resulting raw files to showcase `rawrr`'s functionality across different orbitrap instrument models/generations. In order to find the highest-scoring MS2 scan for `r names(iRTmZ)[1]`++ per file, we now use a simple scoring function, implemented directly in `R` (it counts the number of matching y-ions), instead of running an external search engine (see supplementary file: R script functional test). The web page automatically updates every 30 min using the most recent two files per system as input data. Be aware that the source code is executed in a full parallel fashion (each core processes one raw file) on a Linux server. This shows how scalable analysis pipelines can be constructed starting from basic building blocks (code chunks). It demonstrates that (i) `rawrr`'s data access mechanism works for all types of instrument models and (ii) over network-attached storage.


# Conclusions

Our R package `rawrr` provides direct access to spectral data stored in raw-formatted binary files, thereby eliminating the need for unfavorable conversion to exchange formats. Within the `R` environment, spectral data is presented by using only two non-standard objects representing data items well known to analytical scientists (spectrum & chromatogram). This design choice makes data handling relatively easy and intuitive and requires little knowledge about internal/technical details of the implementation. By using vendor API methods whenever possible, we nevertheless made sure that ease-of-use doesn't impair performance. We also emphasized that our implementation aligns well with common `R` conventions and styles. In the near future, we plan to submit `rawrr` to the Bioconductor project and align further efforts with the R for Mass Spectrometry initiative. In particular, we hope to extend `rawrr` towards the concept of exchangeable backends for data access and parallel computation. These would be necessary next steps towards big computational proteomics in `R`.


## Author contributions

The manuscript was written through contributions of all authors. All authors have given approval to the final version of the manuscript. ‡These authors contributed equally.

<!--
## Funding Sources

Any funds used to support the research of the manuscript should be placed here (per journal style).
-->

## Acknowledgements

We thank Lilly van de Venn for designing the `rawrr` package logo. We are grateful to Jim Shofstahl for providing the `RawFileReader` .NET assembly, `C#` example code, and for answering questions during the development process of `rawrr`.


## Abbreviations

API, advanced programming interface;
BSA, Bovine Serum Albumin;
FTMS, Fourier-transformed mass spectrum;
GUI, graphics user interface;
HRAM, high-resolution accurate mass; 
iRT, indexed retention time;
LC-MS, liquid chromatography followed by mass spectrometry;
OOP, object-oriented programming;
MS, mass spectrometry;
NSI, nanospray ionization;
PRM, parallel reaction monitoring;
TIC, total ion chromatogram;
XIC, extracted ion chromatogram;

## Supporting Information

<!--
Supporting Information - If the manuscript is accompanied by any supporting information for publication, a brief description of the supplementary material is required in the manuscript. The appropriate format is: Supporting Information. Brief statement in nonsentence format listing the contents of the material supplied as Supporting Information.
-->

<!--
- Page 13-14, "Supplements":
If this information is intended as Supporting information, it must be removed and provided as a separate file.   The first page of the file should be a cover page that includes title of the primary article, the authors’ names and affiliations, and a table of contents listing each supporting component by name and brief caption.  Number pages as S-1, S-2, etc
-->

<!--
# Supplements

```{r monoInfo, message=TRUE, echo=TRUE}
rawrr:::.monoInfo()
```

```{r sessionInfo}
sessionInfo()
```
-->

* Manual pages of the package generated by `R CMD Rd2pdf rawrr`.
* R script "functional test": Extended code example generating an pdf report  using [@MSV000086542] and providing the software environment information: Output of `rawrr:::monoInfo()` and `sessionInfo()`.

## Funding Sources

<!--
Funding Sources: Authors are required to report ALL funding sources and grant/award numbers relevant to this manuscript. Enter all sources of funding for ALL authors relevant to this manuscript in BOTH the Open Funder Registry tool in ACS Paragon Plus and in the manuscript to meet this requirement. See http://pubs.acs.org/page/4authors/funder_options.html for complete instructions.
-->

## References
<div id="refs"></div>
